hamilton/function_modifiers/macros.py

```diff
@@ -876,3 +876,134 @@ def optional_config(self) -> Dict[str, Any]:
 #
 #     def __init__(self, *transforms: Applicable, collapse=False):
 #         super(flow, self).__init__(*transforms, collapse=collapse, _chain=False)
+
+
+class post_pipe(base.NodeInjector):
+    def __init__(
+        self,
+        *transforms: Applicable,
+        namespace: NamespaceType = ...,
+        collapse=False,
+        _chain=False,
+    ):
+        """Instantiates a `@post_pipe` decorator.
+
+        :param transforms: step transformations to be applied, in order
+        :param namespace: namespace to apply to all nodes in the pipe. This can be "..." (the default), which resolves to the name of the decorated function, None (which means no namespace), or a string (which means that all nodes will be namespaced with that string). Note that you can either use this *or* namespaces inside pipe()...
+        :param collapse: Whether to collapse this into a single node. This is not currently supported.
+        :param _chain: Whether to chain the first parameter. This is the only mode that is supported. Furthermore, this is not externally exposed. @flow will make use of this.
+        """
+        self.transforms = transforms
+        self.collapse = collapse
+        self.chain = _chain
+        self.namespace = namespace
+
+        if self.collapse:
+            raise NotImplementedError(
+                "Collapsing step() functions as one node is not yet implemented for pipe(). Please reach out if you want this feature."
+            )
+
+        if self.chain:
+            raise NotImplementedError("@flow() is not yet supported -- this is ")
+
+    def transform_dag(
```

Comment:
So I think this indicates that a node injector is not correct -- E.G. you want some variant of a `SubdagTransformer`(which is the correct one). But you're basically doing that by overriding `transform_dag`. 

Your logic appears correct, but I think this might simplify

I think what you want is a `SingleNodeNodeTransformer`, which automatically applies to the sink node. This is akin to `extract_columns`, which applies to the last node in a chain. See this: https://github.com/DAGWorks-Inc/hamilton/blob/b818c15eb66c4d175a48d89087cbe7b62b6c448b/hamilton/function_modifiers/base.py#L557. 

From a graph algebra perspective, what we do is:

1. Take a subdag 
2. Find the "last" node in the subdag (if there are multiple we error out). (this + (1) should be handled by `SingleNodeNodeTransformer`
3. Append one node for each step (this is where we reuse the code from `@pipe` -- you do this with `self.inject_nodes`)
4. Ensure the final node has the name of the original node (you do this as well, nice).
5. Rename the first node (original node) to be `_raw` (you do this as well)

So yes, I believe what you have is correct, but I'm wondering if that class will save you some effort!

Reply:
I had a dilemma about whether to use node injector or node transformer. In this case both seem to describe the desired graph behaviour (either injecting node or transforming the last node). I went with the injector at the end to keep it the same as `pipe` and seemed to me that as long as I respect the API of `transform_subdag` I can override the inner part. 

But looking back I should have just went with the base that has the most fitting functionality. Still on my todo to refactor the common things from `pipe` and `post_pipe` out.

Reply:
So I think `NodeInjector` is problematic because it leaves the first node in. E.G. NodeInject maps node -> list[nodes], then appends. This is (I think) why you implemented `transform_subdag` -- which isn't part of the interface. IMO `SingleNodeNodeTransformer` will likely solve some of the issues for you? Should make the code simpler.